### PR TITLE
TMAP8 SQA

### PR DESCRIPTION
**Pull request by lindsayad**
- [x] test spec information
- [x] documentation

### CODE REVIEW

File: doc/content/source/nodal_kernels/TrappingNodalKernel.md [hunk 1]
```diff
@@ -1,20 +1,53 @@
...
+\begin{equation}
+\frac{\alpha_t}{N} C_t^e C_s
+\end{equation}
+
+where $\alpha_t$ is the trapping rate coefficient, which has dimensions of
+1/time, $N$ is the atomic number density of the host material, $C_t^e$ is the
+concentration of empty trapping sites, and $C_s$ is the concentration of the
+mobile species.
+
+As outlined in [content/index.md#scaling] it is important to scale different
```
Comment (cticenhour) [inline doc/content/source/nodal_kernels/TrappingNodalKernel.md:18]: This is breaking. Try just `[index.md#scaling]` since you've already included the content dir in your config.yml file.

File: test/tests/val-1d/tests [hunk 2]
```diff
@@ -1,14 +1,20 @@
 [Tests]
+  design = 'TrappingNodalKernel.md ReleasingNodalKernel.md'
+  issues = '#12'
   [diffusion_limited]
     type = Exodiff
     input = val-1d-diffusion.i
     exodiff = val-1d-diffusion_out.e
     cli_args = "Mesh/nx=20 Executioner/num_steps=300 Postprocessors/active='' Outputs/out/execute_on='final'"
+    requirement = 'The system shall be able to model a breakthrough problem where diffusion is the rate limiting process.'
+    verification = 'vv-1d.md'
```
Comment (cticenhour) [inline test/tests/val-1d/tests:10]: Typo - `val-1d.md`

File: test/tests/val-1d/tests [hunk 3]
```diff
@@ -1,14 +1,20 @@
...
   []
   [trapping_limited]
     type = Exodiff
     input = val-1d-trapping.i
     exodiff = val-1d-trapping_out.e
     cli_args = 'Executioner/num_steps=10'
+    requirement = 'The system shall be able to model a breakthrough problem where trapping is the rate limiting process.'
+    verification = 'vv-1d.md'
```
Comment (cticenhour) [inline test/tests/val-1d/tests:18]: Typo - `val-1d.md`

**Review by cticenhour — CHANGES_REQUESTED** [review]
This was a quick review just to get the doco passing. I'll do a more thorough one soon!

File: doc/content/source/kernels/LMKernel.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+This NCP function ensures the following conditions:
+
+\begin{align}
+\label{eq:conditions}
+\lambda &\geq 0\\
+u &\geq 0\\
+\lambda u &= 0
+\end{align}
+
+Examining [eq:conditions] it's clear that positive values of $\lambda$ (required
```
[the remark this answers is not in the document]
Comment (lindsayad) [reply doc/content/source/kernels/LMKernel.md:45]: Hmm I know that this syntax works for the `\begin{equation}` environment. I guess that it doesn't work for the `align` environment...

[the remark this answers is not in the document]
Comment (cticenhour) [reply doc/content/source/kernels/LMKernel.md:45]: I had some issues out of `align` in the past (though what they were escapes me at the moment), so I generally use:
```
\begin{equation}
\begin{aligned}
...
\end{aligned}
\end{equation}
```
for anything that needs alignment.

**Comment by lindsayad** [comment]
Oh beautiful pass

**Comment by cticenhour** [comment]
Beautiful! I'll give another review soon 😄

**Comment by cticenhour** [comment]
Seeing this on the current site:
image

Really odd, considering I was seeing it earlier during the failures, and the SQA check passed with flying colors....

**Comment by lindsayad** [comment]
Does this look a little weird too? That's also showing empty shebang links https://mooseframework.inl.gov/sqa/index.html

**Comment by cticenhour** [comment]
The mastodon SQA site (https://mooseframework.inl.gov/mastodon/sqa/index.html) is doing the same thing. @aeslaughter have you seen this yet?

**Comment by lindsayad** [comment]
we've been hacked

**Comment by aeslaughter** [comment]
No. The SQA stuff is being disable by default, thus shows in purple. The SQA stuff has become very slow and requires inspection of the git repo and dowloading sutff from CIVET, etc. 

To enable it, you will need to add --stable to the moosedocs.py build command. This change just made it through MOOSE this week. The plan is to add that flag to the master build of the documentation, but I have not done that yet.

I am going to try that out for a bit and see it that works or if it fails too often work on another solution.

**Comment by cticenhour** [comment]
Thanks @aeslaughter. I will add this flag to my new civet recipe (idaholab/civet_recipes#2793)

**Comment by lindsayad** [comment]
I need to be able to point people to SQA documents today or, from the way people tell it, John Wagner will kick down my door...

File: doc/content/source/kernels/LMKernel.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+\end{equation}
+
+This NCP function ensures the following conditions:
+
+\begin{align}
+\label{eq:conditions}
+\lambda &\geq 0\\
+u &\geq 0\\
+\lambda u &= 0
+\end{align}
```
Comment (cticenhour) [inline doc/content/source/kernels/LMKernel.md:43]: This equation doesn't render correctly. I would stick to using the environment:
```
\begin{equation}
\begin{aligned}
...
\end{aligned}
\end{equation}
```

File: doc/content/source/kernels/BodyForceLM.md [hunk 6]
```diff
@@ -1,20 +1,14 @@
...
-!! Replace these lines with information regarding the BodyForceLM object.
-
-## Example Input File Syntax
-
-!! Describe and include an example of how to use the BodyForceLM object.
+This object is equivalent to [BodyForce.md] except it adds its residual both to
+the primal equation on which the body force is being applied and to a Lagrange
+Multiplier (LM) equation when an LM is present for enforcing non-negative
+concentrations and we are trying to remove the saddle point due to no
+on-diagonal dependence in the LM equation. See [LMKernel.md] for more details.
```
Comment (cticenhour) [inline doc/content/source/kernels/BodyForceLM.md:11]:
```suggestion
This object is equivalent to [BodyForce.md] except it adds its residual both to
the primal equation on which the body force is being applied and to a Lagrange
Multiplier (LM) equation when an LM is present. We use an LM for enforcing non-negative
concentrations and when we are trying to remove the saddle point due to no
on-diagonal dependence in the LM equation. See [LMKernel.md] for more details.
```

This was a bit too long and run-on like for me, so I suggest breaking it up a little.

File: doc/content/source/kernels/CoupledForceLM.md [hunk 7]
```diff
@@ -1,20 +1,14 @@
...
-!! Replace these lines with information regarding the CoupledForceLM object.
-
-## Example Input File Syntax
-
-!! Describe and include an example of how to use the CoupledForceLM object.
+This object is equivalent to [CoupledForce.md] except it adds its residual both to
+the primal equation on which the coupled force is being applied and to a Lagrange
+Multiplier (LM) equation when an LM is present for enforcing non-negative
+concentrations and we are trying to remove the saddle point due to no
+on-diagonal dependence in the LM equation. See [LMKernel.md] for more details.
```
Comment (cticenhour) [inline doc/content/source/kernels/CoupledForceLM.md:11]:
```suggestion
Multiplier (LM) equation when an LM is present. We use an LM for enforcing non-negative
concentrations and when we are trying to remove the saddle point due to no
on-diagonal dependence in the LM equation. See [LMKernel.md] for more details.
```

File: doc/content/source/kernels/LMDiffusion.md [hunk 8]
```diff
@@ -1,20 +1,13 @@
...
 ## Overview
 
-!! Replace these lines with information regarding the LMDiffusion object.
-
-## Example Input File Syntax
-
-!! Describe and include an example of how to use the LMDiffusion object.
+This class adds the strong form of [Diffusion.md] to a Lagrange
+Multiplier (LM) equation when an LM is present for enforcing non-negative
+concentrations and we are trying to remove the saddle point due to no
```
Comment (cticenhour) [inline doc/content/source/kernels/LMDiffusion.md:9]:
```suggestion
concentrations and when we are trying to remove the saddle point due to no
```

File: doc/content/source/kernels/TimeDerivativeLM.md [hunk 9]
```diff
@@ -1,20 +1,14 @@
...
 
-!! Replace these lines with information regarding the TimeDerivativeLM object.
-
-## Example Input File Syntax
-
-!! Describe and include an example of how to use the TimeDerivativeLM object.
+This object is equivalent to [TimeDerivative.md] except it adds its residual both to
+the primal equation containing the time derivative and to a Lagrange
+Multiplier (LM) equation when an LM is present for enforcing non-negative
+concentrations and we are trying to remove the saddle point due to no
```
Comment (cticenhour) [inline doc/content/source/kernels/TimeDerivativeLM.md:10]:
```suggestion
Multiplier (LM) equation when an LM is present. We use an LM for enforcing non-negative
concentrations and when we are trying to remove the saddle point due to no
```

File: doc/content/source/nodal_kernels/TrappingNodalKernel.md [hunk 10]
```diff
@@ -1,20 +1,53 @@
...
+\begin{equation}
+\frac{\alpha_t}{N} C_t^e C_s
+\end{equation}
+
+where $\alpha_t$ is the trapping rate coefficient, which has dimensions of
+1/time, $N$ is the atomic number density of the host material, $C_t^e$ is the
+concentration of empty trapping sites, and $C_s$ is the concentration of the
+mobile species.
+
+As outlined in [index.md#scaling exact=True] it is important to scale different
```
Comment (cticenhour) [inline doc/content/source/nodal_kernels/TrappingNodalKernel.md:18]:
```suggestion
As outlined in [index.md#scaling exact=True], it is important to scale different
```

File: test/tests/ncp-lms/tests [hunk 11]
```diff
@@ -8,26 +10,6 @@
     expect_out = '0 of 22 singular values'
     custom_cmp = nodal.cmp
   []
-  [interpolated_lm_ncp]
```
Comment (cticenhour) [inline test/tests/ncp-lms/tests:11]: Curious - why did you remove these tests (and the associated NCP objects)? Just another way of doing LM that you prefer?

**Review by cticenhour — CHANGES_REQUESTED** [review]
Only small changes requested, and I had a question about the removal of some objects. Looks good overall!

File: doc/content/source/kernels/LMKernel.md (hunk 5 above)
Comment (lindsayad) [reply doc/content/source/kernels/LMKernel.md:43]: image

Hahaha

Comment (cticenhour) [reply doc/content/source/kernels/LMKernel.md:43]: Hahahaha yeah yeah, I know 😂

File: test/tests/ncp-lms/tests (hunk 11 above)
Comment (lindsayad) [reply test/tests/ncp-lms/tests:11]: Since writing these tests a couple years ago (!!), I've gotten a lot more experience with NCP-type constraints in mechanical contact and the literature suggests performing the NCP constraints on essentially a per-dof basis. So if the LM is Lagrange, then the NCP constraint should be enforced on a per-node basis (which I do with `NodalKernels`). 

Just within the last week @jbadger95 has seen oscillatory behavior when trying to set an LM to zero in an "elemental" way as opposed to on a per-dof basis. You can imagine for example that if you do an integrated/elemental approach, an allowed solution to, say, \int_{elem} \lambda dV using a two point quadrature rule would be for the solution at qp1 to be +C and the solution at qp2 to be -C. We want to avoid these oscillatory solutions

Comment (cticenhour) [reply test/tests/ncp-lms/tests:11]: I understand! Thanks for the details :)

**Review by cticenhour — APPROVED** [review]
Beautiful code. Beautiful doco.
